pkg/virt-controller/watch/vmi.go

```diff
@@ -861,6 +867,50 @@ func (c *VMIController) hotplugPodsReady(vmi *virtv1.VirtualMachineInstance, vir
 	return true, nil
 }
 
+func hasMaintenancePodPVC(guestfsPods []*k8sv1.Pod, pvc string) bool {
+	for _, pod := range guestfsPods {
+		if pod.ObjectMeta.Labels[guestfs.VirtGuestfsPVCLabel] == pvc {
+			return true
+		}
+	}
+	return false
+}
+
+// checkMaintenancePodActiveOnVolumes checks if one of the volumes of the VMI is in used by a pod created by virtctl guestfs
+func (c *VMIController) checkMaintenancePodActiveOnVolumes(vmi *virtv1.VirtualMachineInstance) error {
+	objs, err := c.podInformer.GetIndexer().ByIndex(cache.NamespaceIndex, vmi.Namespace)
+	if err != nil {
+		return err
+	}
+	var guestfsPods []*k8sv1.Pod
+	for _, obj := range objs {
+		pod := obj.(*k8sv1.Pod)
+		if pod.ObjectMeta.Labels[virtv1.AppLabel] == guestfs.VirtGuestfsName {
+			guestfsPods = append(guestfsPods, pod)
+		}
+	}
+
+	// No guestfs pods
+	if len(guestfsPods) <= 0 {
+		return nil
+	}
+
+	for _, v := range vmi.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil || v.DataVolume != nil {
+			var claimName string
+			if v.PersistentVolumeClaim != nil {
+				claimName = v.PersistentVolumeClaim.ClaimName
+			} else {
+				claimName = v.DataVolume.Name
+			}
+			if hasMaintenancePodPVC(guestfsPods, claimName) {
+				return fmt.Errorf("volume %s is in use by virtctl guestfs", claimName)
+			}
+		}
```

Comment:
`n^2` operation. I think we can get it down to 'n', see suggestion above

Reply:
Done